File: py/app/custom_preset.py
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Dict

from app import custom_preset_pb2
from google.protobuf.json_format import MessageToJson, Parse

class PlotType(Enum):
    UNKNOWN = 0
    LINE = 1
    SCATTER = 2


@dataclass
class Size:
    w: float = 0.0
    h: float = 0.0
    
    _proto_message: custom_preset_pb2.Size = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = custom_preset_pb2.Size()

    def _populate_proto(self):
        self._proto_message.w = self.w
        self._proto_message.h = self.h

    @classmethod
    def _from_proto(cls, proto_message) -> "Size":
        return cls(
            w = proto_message.w,
            h = proto_message.h,
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "Size":
        message = custom_preset_pb2.Size()
        message.ParseFromString(data)
        return cls(
            w = message.w,
            h = message.h,
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "Size":
        message = custom_preset_pb2.Size()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

@dataclass
class GraphElement:
    label: str = ""
    type: PlotType = None
    xField: str = ""
    yField: str = ""
    color: int = 0
    axisX: int = 0
    axisY: int = 0
    
    _proto_message: custom_preset_pb2.GraphElement = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = custom_preset_pb2.GraphElement()

    def _populate_proto(self):
        self._proto_message.label = self.label
        self._proto_message.type = self.type.value
        self._proto_message.xField = self.xField
        self._proto_message.yField = self.yField
        self._proto_message.color = self.color
        self._proto_message.axisX = self.axisX
        self._proto_message.axisY = self.axisY

    @classmethod
    def _from_proto(cls, proto_message) -> "GraphElement":
        return cls(
            label = proto_message.label,
            type = PlotType(proto_message.type),
            xField = proto_message.xField,
            yField = proto_message.yField,
            color = proto_message.color,
            axisX = proto_message.axisX,
            axisY = proto_message.axisY,
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "GraphElement":
        message = custom_preset_pb2.GraphElement()
        message.ParseFromString(data)
        return cls(
            label = message.label,
            type = PlotType(message.type),
            xField = message.xField,
            yField = message.yField,
            color = message.color,
            axisX = message.axisX,
            axisY = message.axisY,
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "GraphElement":
        message = custom_preset_pb2.GraphElement()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

@dataclass
class Graph:
    id: str = ""
    title: str = ""
    tags: List[str] = field(default_factory=list)
    signals: List[GraphElement] = field(default_factory=list)
    otherElements: List[str] = field(default_factory=list)
    hasTooltip: bool = False
    author: str = ""
    implotFlags: int = 0
    axisFlags: int = 0
    
    _proto_message: custom_preset_pb2.Graph = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = custom_preset_pb2.Graph()

    def _populate_proto(self):
        self._proto_message.id = self.id
        self._proto_message.title = self.title
        del self._proto_message.tags[:]
        for val in self.tags:
            self._proto_message.tags.append(val)
        del self._proto_message.signals[:]
        for val in self.signals:
            val._populate_proto()
            tmp = self._proto_message.signals.add()
            tmp.CopyFrom(val._proto_message)
        del self._proto_message.otherElements[:]
        for val in self.otherElements:
            self._proto_message.otherElements.append(val)
        self._proto_message.hasTooltip = self.hasTooltip
        self._proto_message.author = self.author
        self._proto_message.implotFlags = self.implotFlags
        self._proto_message.axisFlags = self.axisFlags

    @classmethod
    def _from_proto(cls, proto_message) -> "Graph":
        return cls(
            id = proto_message.id,
            title = proto_message.title,
            tags=[str(val) for val in proto_message.tags],
            signals=[GraphElement._from_proto(val) for val in proto_message.signals],
            otherElements=[str(val) for val in proto_message.otherElements],
            hasTooltip = proto_message.hasTooltip,
            author = proto_message.author,
            implotFlags = proto_message.implotFlags,
            axisFlags = proto_message.axisFlags,
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "Graph":
        message = custom_preset_pb2.Graph()
        message.ParseFromString(data)
        return cls(
            id = message.id,
            title = message.title,
            tags = [str(val) for val in message.tags],
            signals = [GraphElement._from_proto(val) for val in message.signals],
            otherElements = [str(val) for val in message.otherElements],
            hasTooltip = message.hasTooltip,
            author = message.author,
            implotFlags = message.implotFlags,
            axisFlags = message.axisFlags,
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "Graph":
        message = custom_preset_pb2.Graph()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

@dataclass
class GridPlot:
    name: str = ""
    id: str = ""
    size: Size = None
    flags: int = 0
    graphID: List[str] = field(default_factory=list)
    
    _proto_message: custom_preset_pb2.GridPlot = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = custom_preset_pb2.GridPlot()

    def _populate_proto(self):
        self._proto_message.name = self.name
        self._proto_message.id = self.id
        if self.size:
            self.size._populate_proto()
            self._proto_message.size.CopyFrom(self.size._proto_message)
        self._proto_message.flags = self.flags
        del self._proto_message.graphID[:]
        for val in self.graphID:
            self._proto_message.graphID.append(val)

    @classmethod
    def _from_proto(cls, proto_message) -> "GridPlot":
        return cls(
            name = proto_message.name,
            id = proto_message.id,
            size = Size._from_proto(proto_message.size),
            flags = proto_message.flags,
            graphID=[str(val) for val in proto_message.graphID],
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "GridPlot":
        message = custom_preset_pb2.GridPlot()
        message.ParseFromString(data)
        return cls(
            name = message.name,
            id = message.id,
            size = (
                Size._from_proto(message.size)
                if message.HasField("size")
                else None
            ),
            flags = message.flags,
            graphID = [str(val) for val in message.graphID],
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "GridPlot":
        message = custom_preset_pb2.GridPlot()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

@dataclass
class CustomPage:
    id: str = ""
    name: str = ""
    size: Size = None
    author: str = ""
    gridCells: List[int] = field(default_factory=list)
    gridPlotID: List[str] = field(default_factory=list)
    
    _proto_message: custom_preset_pb2.CustomPage = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = custom_preset_pb2.CustomPage()

    def _populate_proto(self):
        self._proto_message.id = self.id
        self._proto_message.name = self.name
        if self.size:
            self.size._populate_proto()
            self._proto_message.size.CopyFrom(self.size._proto_message)
        self._proto_message.author = self.author
        del self._proto_message.gridCells[:]
        for val in self.gridCells:
            self._proto_message.gridCells.append(val)
        del self._proto_message.gridPlotID[:]
        for val in self.gridPlotID:
            self._proto_message.gridPlotID.append(val)

    @classmethod
    def _from_proto(cls, proto_message) -> "CustomPage":
        return cls(
            id = proto_message.id,
            name = proto_message.name,
            size = Size._from_proto(proto_message.size),
            author = proto_message.author,
            gridCells=[int(val) for val in proto_message.gridCells],
            gridPlotID=[str(val) for val in proto_message.gridPlotID],
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "CustomPage":
        message = custom_preset_pb2.CustomPage()
        message.ParseFromString(data)
        return cls(
            id = message.id,
            name = message.name,
            size = (
                Size._from_proto(message.size)
                if message.HasField("size")
                else None
            ),
            author = message.author,
            gridCells = [int(val) for val in message.gridCells],
            gridPlotID = [str(val) for val in message.gridPlotID],
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "CustomPage":
        message = custom_preset_pb2.CustomPage()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

@dataclass
class PresetMappings:
    plot_configs: Dict[str, Graph] = field(default_factory=dict)
    subplot_configs: Dict[str, GridPlot] = field(default_factory=dict)
    preset_configs: Dict[str, CustomPage] = field(default_factory=dict)
    
    _proto_message: custom_preset_pb2.PresetMappings = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = custom_preset_pb2.PresetMappings()

    def _populate_proto(self):
        self._proto_message.plot_configs.clear()
        for key, val in self.plot_configs.items():
            val._populate_proto()
            tmp = self._proto_message.plot_configs.setdefault(key)
            tmp.CopyFrom(val._proto_message)
        self._proto_message.subplot_configs.clear()
        for key, val in self.subplot_configs.items():
            val._populate_proto()
            tmp = self._proto_message.subplot_configs.setdefault(key)
            tmp.CopyFrom(val._proto_message)
        self._proto_message.preset_configs.clear()
        for key, val in self.preset_configs.items():
            val._populate_proto()
            tmp = self._proto_message.preset_configs.setdefault(key)
            tmp.CopyFrom(val._proto_message)

    @classmethod
    def _from_proto(cls, proto_message) -> "PresetMappings":
        return cls(
            plot_configs={key: Graph._from_proto(val) for key, val in proto_message.plot_configs.items()},
            subplot_configs={key: GridPlot._from_proto(val) for key, val in proto_message.subplot_configs.items()},
            preset_configs={key: CustomPage._from_proto(val) for key, val in proto_message.preset_configs.items()},
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "PresetMappings":
        message = custom_preset_pb2.PresetMappings()
        message.ParseFromString(data)
        return cls(
            plot_configs = {key: Graph._from_proto(val) for key, val in message.plot_configs.items()},
            subplot_configs = {key: GridPlot._from_proto(val) for key, val in message.subplot_configs.items()},
            preset_configs = {key: CustomPage._from_proto(val) for key, val in message.preset_configs.items()},
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "PresetMappings":
        message = custom_preset_pb2.PresetMappings()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

File: src/app/custom_preset.cpp
#include "app/custom_preset.h"

#include <google/protobuf/util/json_util.h>
#include <tuple>

namespace Serializers
{
namespace App
{
Size::Size(const PbApp::Size& protobuf) {
    w = protobuf.w();
    h = protobuf.h();
}

Size::operator PbApp::Size() const {
    PbApp::Size ret;
    ret.set_w(w);
    ret.set_h(h);
    return ret;
}

std::string Size::serializeAsJsonString() const {
    PbApp::Size protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string Size::serializeAsProtobufString() const {
    PbApp::Size protobuf(*this);
    return protobuf.SerializeAsString();
}

bool Size::deserializeFromJsonString(const std::string& str) {
    PbApp::Size protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool Size::deserializeFromProtobufString(const std::string& str) {
    PbApp::Size protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

GraphElement::GraphElement(const PbApp::GraphElement& protobuf) {
    label = protobuf.label();
    type = static_cast<PlotType>(protobuf.type());
    xField = protobuf.xfield();
    yField = protobuf.yfield();
    color = protobuf.color();
    axisX = protobuf.axisx();
    axisY = protobuf.axisy();
}

GraphElement::operator PbApp::GraphElement() const {
    PbApp::GraphElement ret;
    ret.set_label(label);
    ret.set_type(static_cast<PbApp::PlotType>(type));
    ret.set_xfield(xField);
    ret.set_yfield(yField);
    ret.set_color(color);
    ret.set_axisx(axisX);
    ret.set_axisy(axisY);
    return ret;
}

std::string GraphElement::serializeAsJsonString() const {
    PbApp::GraphElement protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string GraphElement::serializeAsProtobufString() const {
    PbApp::GraphElement protobuf(*this);
    return protobuf.SerializeAsString();
}

bool GraphElement::deserializeFromJsonString(const std::string& str) {
    PbApp::GraphElement protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool GraphElement::deserializeFromProtobufString(const std::string& str) {
    PbApp::GraphElement protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

Graph::Graph(const PbApp::Graph& protobuf) {
    id = protobuf.id();
    title = protobuf.title();
    tags = {protobuf.tags().begin(), protobuf.tags().end()};
    signals = {protobuf.signals().begin(), protobuf.signals().end()};
    otherElements = {protobuf.otherelements().begin(), protobuf.otherelements().end()};
    hasTooltip = protobuf.hastooltip();
    author = protobuf.author();
    implotFlags = protobuf.implotflags();
    axisFlags = protobuf.axisflags();
}

Graph::operator PbApp::Graph() const {
    PbApp::Graph ret;
    ret.set_id(id);
    ret.set_title(title);
    *(ret.mutable_tags()) = {tags.begin(), tags.end()};
    *(ret.mutable_signals()) = {signals.begin(), signals.end()};
    *(ret.mutable_otherelements()) = {otherElements.begin(), otherElements.end()};
    ret.set_hastooltip(hasTooltip);
    ret.set_author(author);
    ret.set_implotflags(implotFlags);
    ret.set_axisflags(axisFlags);
    return ret;
}

std::string Graph::serializeAsJsonString() const {
    PbApp::Graph protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string Graph::serializeAsProtobufString() const {
    PbApp::Graph protobuf(*this);
    return protobuf.SerializeAsString();
}

bool Graph::deserializeFromJsonString(const std::string& str) {
    PbApp::Graph protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool Graph::deserializeFromProtobufString(const std::string& str) {
    PbApp::Graph protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

GridPlot::GridPlot(const PbApp::GridPlot& protobuf) {
    name = protobuf.name();
    id = protobuf.id();
    size = protobuf.size();
    flags = protobuf.flags();
    graphID = {protobuf.graphid().begin(), protobuf.graphid().end()};
}

GridPlot::operator PbApp::GridPlot() const {
    PbApp::GridPlot ret;
    ret.set_name(name);
    ret.set_id(id);
    *(ret.mutable_size()) = size;
    ret.set_flags(flags);
    *(ret.mutable_graphid()) = {graphID.begin(), graphID.end()};
    return ret;
}

std::string GridPlot::serializeAsJsonString() const {
    PbApp::GridPlot protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string GridPlot::serializeAsProtobufString() const {
    PbApp::GridPlot protobuf(*this);
    return protobuf.SerializeAsString();
}

bool GridPlot::deserializeFromJsonString(const std::string& str) {
    PbApp::GridPlot protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool GridPlot::deserializeFromProtobufString(const std::string& str) {
    PbApp::GridPlot protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

CustomPage::CustomPage(const PbApp::CustomPage& protobuf) {
    id = protobuf.id();
    name = protobuf.name();
    size = protobuf.size();
    author = protobuf.author();
    gridCells = {protobuf.gridcells().begin(), protobuf.gridcells().end()};
    gridPlotID = {protobuf.gridplotid().begin(), protobuf.gridplotid().end()};
}

CustomPage::operator PbApp::CustomPage() const {
    PbApp::CustomPage ret;
    ret.set_id(id);
    ret.set_name(name);
    *(ret.mutable_size()) = size;
    ret.set_author(author);
    *(ret.mutable_gridcells()) = {gridCells.begin(), gridCells.end()};
    *(ret.mutable_gridplotid()) = {gridPlotID.begin(), gridPlotID.end()};
    return ret;
}

std::string CustomPage::serializeAsJsonString() const {
    PbApp::CustomPage protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string CustomPage::serializeAsProtobufString() const {
    PbApp::CustomPage protobuf(*this);
    return protobuf.SerializeAsString();
}

bool CustomPage::deserializeFromJsonString(const std::string& str) {
    PbApp::CustomPage protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool CustomPage::deserializeFromProtobufString(const std::string& str) {
    PbApp::CustomPage protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

PresetMappings::PresetMappings(const PbApp::PresetMappings& protobuf) {
    plot_configs = {protobuf.plot_configs().begin(), protobuf.plot_configs().end()};
    subplot_configs = {protobuf.subplot_configs().begin(), protobuf.subplot_configs().end()};
    preset_configs = {protobuf.preset_configs().begin(), protobuf.preset_configs().end()};
}

PresetMappings::operator PbApp::PresetMappings() const {
    PbApp::PresetMappings ret;
    *(ret.mutable_plot_configs()) = {plot_configs.begin(), plot_configs.end()};
    *(ret.mutable_subplot_configs()) = {subplot_configs.begin(), subplot_configs.end()};
    *(ret.mutable_preset_configs()) = {preset_configs.begin(), preset_configs.end()};
    return ret;
}

std::string PresetMappings::serializeAsJsonString() const {
    PbApp::PresetMappings protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string PresetMappings::serializeAsProtobufString() const {
    PbApp::PresetMappings protobuf(*this);
    return protobuf.SerializeAsString();
}

bool PresetMappings::deserializeFromJsonString(const std::string& str) {
    PbApp::PresetMappings protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool PresetMappings::deserializeFromProtobufString(const std::string& str) {
    PbApp::PresetMappings protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}
}
}